appmode/server_extension.py

```diff
@@ -87,52 +99,62 @@ def get(self, path):
 
     #===========================================================================
     @web.authenticated
-    @gen.coroutine
-    def post(self, path):
+    async def post(self, path):
         assert self.get_body_arguments("appmode_action")[0] == "delete"
         path = path.strip('/')
         self.log.info('Appmode deleting: %s', path)
 
         # delete session, including the kernel
         sm = self.session_manager
-        if gen.is_coroutine_function(sm.get_session):
-            s = yield sm.get_session(path=path)
-        else:
-            s = sm.get_session(path=path)
-        if gen.is_coroutine_function(sm.delete_session):
-            yield sm.delete_session(session_id=s['id'])
-        else:
-            sm.delete_session(session_id=s['id'])
+
+        s = sm.get_session(path=path)
+        if isinstance(s, collections.abc.Awaitable):
+            s = await s
+        sd = sm.delete_session(session_id=s['id'])
+        if isinstance(sd, collections.abc.Awaitable):
+            await sd
 
         # delete tmp copy
         cm = self.contents_manager
-        cm.delete(path)
-        self.finish()
+        pd = cm.delete(path)
+        if isinstance(pd, collections.abc.Awaitable):
+            await pd
+        await self.finish()
 
     #===========================================================================
-    def mk_tmp_copy(self, path):
+    async def mk_tmp_copy(self, path):
         cm = self.contents_manager
-
         # find tmp_path
-        dirname = os.path.dirname(path)
+        dirname = self.temp_dir or os.path.dirname(path)
+        if dirname and not os.path.exists(dirname):
+            os.makedirs(dirname)
         fullbasename = os.path.basename(path)
         basename, ext = os.path.splitext(fullbasename)
+        if self.hidden_temp_files:
+            basename = '.' + basename
         for i in itertools.count():
-            tmp_path = "%s/.%s-%i%s"%(dirname, basename, i, ext)
+            tmp_path = "%s/%s-%i%s"%(dirname, basename, i, ext)
             if not cm.exists(tmp_path):
                 break
 
         # create tmp copy - allows opening same notebook multiple times
         self.log.info("Appmode creating tmp copy: "+tmp_path)
-        cm.copy(path, tmp_path)
-
-        return(tmp_path)
+        pc = cm.copy(path, tmp_path)
+        if isinstance(pc, collections.abc.Awaitable):
```

Comment:
Nice check here! I like it.

Reply:
Not sure it is easy to implement this pattern for all such changes. You use a lot of

```python
        pd = cm.delete(path)
        if isinstance(pd, collections.abc.Awaitable):
            await pd
```

Is that possible to add this to the `self.context_manager`?

Reply:
The content_manager/session_manager is part of the underlying jupyter stack I believe. These checks are repetitive, but they are similar to what was there before, i.e 
```
        if gen.is_coroutine_function(sm.get_session):
            s = yield sm.get_session(path=path)
        else:
            s = sm.get_session(path=path)
        if gen.is_coroutine_function(sm.delete_session):
            yield sm.delete_session(session_id=s['id'])
        else:
            sm.delete_session(session_id=s['id'])
```
The session manager and content manager seem to have moved across to using the newer native async functionality instead of the older tornado coroutines, hence the change to checks on async. There may be a better or more canonical way to do these checks, I haven't seen any other examples, this is just what I came up with to do the same thing as the previous code was trying to do.

We could perhaps make a wrapper for it, like
```
def check_async(x):
    if isinstance(x, collections.abc.Awaitable):
        x = await x
    return x
````
Calls then become ```check_async(cm.delete(path))``` or ```s = check_async(sm.get_session(path=path))```. I can swap it over to doing that if you think it would be better?

Since these checks are for backwards compatibility, another option would be to remove them and just use the awaits, i.e.
```
s = await sm.get_session(path=path)
```
Would then need to version pin the dependency that changes over to async content and session managers.

Reply:
Following the pattern in jupyter_core for handling functions that may or may not be async (see https://github.com/jupyter/jupyter_core/blob/main/jupyter_core/utils/__init__.py#L172), I've added a simple ensure_async function to wrap the contents manager and session manager calls.

This cleans up the repetitive checks for Awaitable objects, and I believe resolves this issue.